includes/Discovery.php

```diff
@@ -231,6 +245,56 @@ protected function get_schemaorg_metadata(): array {
 		return apply_filters( 'web_stories_story_schema_metadata', $metadata, $post );
 	}
 
+	/**
+	 * Get product schema data.
+	 *
+	 * @since 1.22.0
+	 *
+	 * @param Story $story Story object.
+	 * @return array
+	 */
+	protected function get_product_data( Story $story ): array {
+		/**
+		 * Product data.
+		 *
+		 * @var array|false $products
+		 */
+		$products = get_post_meta( $story->get_id(), $this->product_meta::PRODUCTS_POST_META_KEY, true );
+		if ( ! \is_array( $products ) ) {
+			return [];
+		}
+		$product_data = [];
+		foreach ( $products as $product ) {
+			$data = [
+				'@type'     => 'Product',
+				'brand'     => $product['productBrand'] ?? '',
+				'productID' => $product['productId'] ?? '',
+				'url'       => $product['productUrl'] ?? '',
+				'name'      => $product['productTitle'] ?? '',
+			];
+			if ( isset( $product['productImages'] ) && $product['productImages'] ) {
+				$data['image'] = $product['productImages'][0]['url'];
+			}
+			if ( isset( $product['aggregateRating'] ) && $product['aggregateRating'] ) {
+				$data['aggregateRating'] = [
+					'@type'       => 'AggregateRating',
+					'ratingValue' => $product['aggregateRating']['ratingValue'] ?? 0,
+					'reviewCount' => $product['aggregateRating']['reviewCount'] ?? 0,
+					'url'         => $product['aggregateRating']['reviewUrl'] ?? '',
+				];
+			}
+			$product_data[] = $data;
+		}
+
+		return [
+			'products' => [
```

Comment:
```suggestion
			'mainEntity' => [
```

Otherwise it does not pass the schema markup validator.